core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,430 @@
+
+                if (partitionToRebalance == null) {
+                    return false;
+                }
+
+                PartitionInfo partitionInfo = partitionInfos[partitionToRebalance.id];
+
+                // If partition is already balanced or min skewed writer is already writing to that partition, then skip
+                // this partition and move on to the other partition inside max writer.
+                if (isPartitionRebalanced(partitionToRebalance.id) || partitionInfo.containsWriter(to.id)) {
+                    continue;
+                }
+
+                long estimatedPartitionWrittenBytesSinceLastRebalance = estimatePartitionWrittenBytesSinceLastRebalance(from, partitionToRebalance.rowCount);
+                long estimatedPartitionWrittenBytes = estimatedPartitionWrittenBytesSinceLastRebalance + partitionInfo.getPhysicalWrittenBytesAtLastRebalancePerWriter();
+
+                // Scale the partition when estimated physicalWrittenBytes is greater than writerMinSize.
+                if (partitionInfo.getWriterCount() <= numberOfWriters && estimatedPartitionWrittenBytes >= writerMinSize) {
+                    partitionInfo.addWriter(to.id);
+                    rebalancedPartitions.add(partitionToRebalance.id);
+                    updateWriterPriorities(from, to, estimatedPartitionWrittenBytesSinceLastRebalance);
+                    log.debug(format("Scaled partition (%s) to writer %s with writer count %s", partitionToRebalance.id, to.id, partitionInfo.getWriterCount()));
+                    return true;
+                }
+
+                return false;
+            }
+        }
+
+        private void updateWriterPriorities(WriterId from, WriterId to, long estimatedPartitionWrittenBytesSinceLastRebalance)
+        {
+            // Since a partition is rebalanced from max to min skewed writer, decrease the priority of max
+            // writer as well as increase the priority of min writer.
+            writerPriorities[from.id] -= (estimatedPartitionWrittenBytesSinceLastRebalance / 2);
+            writerPriorities[to.id] += (estimatedPartitionWrittenBytesSinceLastRebalance / 2);
+        }
+
+        private long getWriterPriority(WriterId writer)
+        {
+            return writerPriorities[writer.id];
+        }
+
+        private boolean isPartitionRebalanced(int partitionId)
+        {
+            return rebalancedPartitions.contains(partitionId);
+        }
+
+        private long estimatePartitionWrittenBytesSinceLastRebalance(WriterId writer, long partitionRowCount)
+        {
+            return (writerPhysicalWrittenBytesSinceLastRebalance[writer.id] * partitionRowCount) / writerRowCountSinceLastRebalance[writer.id];
+        }
+    }
+
+    @ThreadSafe
+    private static class PartitionInfo
+    {
+        private final List<Integer> writerAssignments;
+        private final AtomicLong physicalWrittenBytesAtLastRebalance = new AtomicLong(0);
```

Comment:
I would add a comment that it partition rebalance and not about `rebalancePartitions` call